File: apps/ios/Pathfinding/Pathfinding/Models/WiFi.swift
import Foundation

// MARK: - WiFi Spot

/// Public WiFi hotspot information
struct WiFiSpot: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let nameEn: String?
    let type: WiFiSpotType
    let cityId: String
    let address: String?
    let latitude: Double
    let longitude: Double
    let ssid: String?
    let requiresPassword: Bool
    let isFree: Bool
    let speedMbps: Double?
    let openingHours: String?
    let averageRating: Double
    let ratingCount: Int
    let description: String?
    let imageUrls: [String]?
    let poiId: String?
    let isVerified: Bool
    let submittedBy: String
    let createdAt: String
    let updatedAt: String
    let distance: Double? // For nearby queries

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case nameEn = "name_en"
        case type
        case cityId = "city_id"
        case address
        case latitude
        case longitude
        case ssid
        case requiresPassword = "requires_password"
        case isFree = "is_free"
        case speedMbps = "speed_mbps"
        case openingHours = "opening_hours"
        case averageRating = "average_rating"
        case ratingCount = "rating_count"
        case description
        case imageUrls = "image_urls"
        case poiId = "poi_id"
        case isVerified = "is_verified"
        case submittedBy = "submitted_by"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case distance
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: WiFiSpot, rhs: WiFiSpot) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - WiFi Spot Type

enum WiFiSpotType: String, Codable, CaseIterable {
    case hotel
    case restaurant
    case cafe
    case airport
    case trainStation = "train_station"
    case shoppingMall = "shopping_mall"
    case library
    case coworking
    case `public`
    case other

    var displayName: String {
        switch self {
        case .hotel: return "Hotel"
        case .restaurant: return "Restaurant"
        case .cafe: return "Cafe"
        case .airport: return "Airport"
        case .trainStation: return "Train Station"
        case .shoppingMall: return "Shopping Mall"
        case .library: return "Library"
        case .coworking: return "Coworking"
        case .public: return "Public"
        case .other: return "Other"
        }
    }

    var displayNameChinese: String {
        switch self {
        case .hotel: return "酒店"
        case .restaurant: return "餐厅"
        case .cafe: return "咖啡厅"
        case .airport: return "机场"
        case .trainStation: return "火车站"
        case .shoppingMall: return "购物中心"
        case .library: return "图书馆"
        case .coworking: return "共享办公"
        case .public: return "公共"
        case .other: return "其他"
        }
    }

    var icon: String {
        switch self {
        case .hotel: return "bed.double.fill"
        case .restaurant: return "fork.knife"
        case .cafe: return "cup.and.saucer.fill"
        case .airport: return "airplane"
        case .trainStation: return "tram.fill"
        case .shoppingMall: return "bag.fill"
        case .library: return "books.vertical.fill"
        case .coworking: return "desktopcomputer"
        case .public: return "wifi"
        case .other: return "mappin.circle.fill"
        }
    }
}

// MARK: - WiFi Credential

/// User-saved WiFi password
struct WiFiCredential: Codable, Identifiable, Hashable {
    let id: String
    let userId: String
    let wifiSpotId: String?
    let name: String
    let ssid: String
    let password: String
    let securityType: SecurityType?
    let locationName: String?
    let latitude: Double?
    let longitude: Double?
    let notes: String?
    let isShared: Bool
    let lastUsedAt: String?
    let createdAt: String
    let updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case wifiSpotId = "wifi_spot_id"
        case name
        case ssid
        case password
        case securityType = "security_type"
        case locationName = "location_name"
        case latitude
        case longitude
        case notes
        case isShared = "is_shared"
        case lastUsedAt = "last_used_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: WiFiCredential, rhs: WiFiCredential) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - Security Type

enum SecurityType: String, Codable, CaseIterable {
    case open
    case wep
    case wpa
    case wpa2
    case wpa3
    case unknown

    var displayName: String {
        switch self {
        case .open: return "Open"
        case .wep: return "WEP"
        case .wpa: return "WPA"
        case .wpa2: return "WPA2"
        case .wpa3: return "WPA3"
        case .unknown: return "Unknown"
        }
    }
}

// MARK: - WiFi Review

/// User quality rating for a WiFi spot
struct WiFiReview: Codable, Identifiable, Hashable {
    let id: String
    let userId: String
    let wifiSpotId: String
    let speedRating: Int
    let stabilityRating: Int
    let easeOfAccessRating: Int
    let overallRating: Int
    let comment: String?
    let speedTestResult: Double?
    let connectionTime: String?
    let deviceType: String?
    let visitDate: String?
    let helpfulCount: Int
    let createdAt: String
    let updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case wifiSpotId = "wifi_spot_id"
        case speedRating = "speed_rating"
        case stabilityRating = "stability_rating"
        case easeOfAccessRating = "ease_of_access_rating"
        case overallRating = "overall_rating"
        case comment
        case speedTestResult = "speed_test_result"
        case connectionTime = "connection_time"
        case deviceType = "device_type"
        case visitDate = "visit_date"
        case helpfulCount = "helpful_count"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    // MARK: - Hashable

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: WiFiReview, rhs: WiFiReview) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - API Response Wrappers

struct WiFiSpotListResponse: Codable {
    let data: [WiFiSpot]
}

struct WiFiSpotResponse: Codable {
    let data: WiFiSpot
}

struct WiFiCredentialListResponse: Codable {
    let data: [WiFiCredential]
}

struct WiFiCredentialResponse: Codable {
    let data: WiFiCredential
}

struct WiFiReviewListResponse: Codable {
    let data: [WiFiReview]
}

struct WiFiReviewResponse: Codable {
    let data: WiFiReview
}

struct WiFiCreateResponse: Codable {
    let data: WiFiCreateData
}

struct WiFiCreateData: Codable {
    let id: String
}

// MARK: - Request Bodies

struct CreateWiFiSpotRequest: Codable {
    let name: String
    let nameEn: String?
    let type: String
    let cityId: String
    let address: String?
    let latitude: Double
    let longitude: Double
    let ssid: String?
    let requiresPassword: Bool
    let isFree: Bool
    let speedMbps: Double?
    let openingHours: String?
    let description: String?
    let imageUrls: [String]?
    let poiId: String?

    enum CodingKeys: String, CodingKey {
        case name
        case nameEn = "name_en"
        case type
        case cityId = "city_id"
        case address
        case latitude
        case longitude
        case ssid
        case requiresPassword = "requires_password"
        case isFree = "is_free"
        case speedMbps = "speed_mbps"
        case openingHours = "opening_hours"
        case description
        case imageUrls = "image_urls"
        case poiId = "poi_id"
    }
}

struct CreateWiFiCredentialRequest: Codable {
    let wifiSpotId: String?
    let name: String
    let ssid: String
    let password: String
    let securityType: String?
    let locationName: String?
    let latitude: Double?
    let longitude: Double?
    let notes: String?
    let isShared: Bool?

    enum CodingKeys: String, CodingKey {
        case wifiSpotId = "wifi_spot_id"
        case name
        case ssid
        case password
        case securityType = "security_type"
        case locationName = "location_name"
        case latitude
        case longitude
        case notes
        case isShared = "is_shared"
    }
}

struct CreateWiFiReviewRequest: Codable {
    let wifiSpotId: String
    let speedRating: Int
    let stabilityRating: Int
    let easeOfAccessRating: Int
    let overallRating: Int
    let comment: String?
    let speedTestResult: Double?
    let connectionTime: String?
    let deviceType: String?
    let visitDate: String?

    enum CodingKeys: String, CodingKey {
        case wifiSpotId = "wifi_spot_id"
        case speedRating = "speed_rating"
        case stabilityRating = "stability_rating"
        case easeOfAccessRating = "ease_of_access_rating"
        case overallRating = "overall_rating"
        case comment
        case speedTestResult = "speed_test_result"
        case connectionTime = "connection_time"
        case deviceType = "device_type"
        case visitDate = "visit_date"
    }
}

File: apps/api/src/routes/wifi.ts
/**
 * WiFi Routes - API endpoints for WiFi spots, credentials, and reviews
 */

import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import type { AuthContext } from '../middleware/auth';
import {
  WifiSpotService,
  WifiCredentialService,
  WifiReviewService,
} from '../services/wifiService';

// ============================================
// Validators
// ============================================

const wifiSpotTypeSchema = z.enum([
  'hotel',
  'restaurant',
  'cafe',
  'airport',
  'train_station',
  'shopping_mall',
  'library',
  'coworking',
  'public',
  'other',
]);

const securityTypeSchema = z.enum([
  'open',
  'wep',
  'wpa',
  'wpa2',
  'wpa3',
  'unknown',
]);

const listSpotsQuerySchema = z.object({
  cityId: z.string().optional(),
  type: wifiSpotTypeSchema.optional(),
  limit: z.coerce.number().min(1).max(100).optional(),
});

const nearbySpotsQuerySchema = z.object({
  latitude: z.coerce.number().min(-90).max(90),
  longitude: z.coerce.number().min(-180).max(180),
  radiusKm: z.coerce.number().min(0.1).max(50).default(5),
  type: wifiSpotTypeSchema.optional(),
  limit: z.coerce.number().min(1).max(100).optional(),
});

const searchSpotsQuerySchema = z.object({
  query: z.string().min(1),
  cityId: z.string().optional(),
  type: wifiSpotTypeSchema.optional(),
  limit: z.coerce.number().min(1).max(100).optional(),
});

const createSpotBodySchema = z.object({
  name: z.string().min(1).max(200),
  nameEn: z.string().max(200).optional(),
  type: wifiSpotTypeSchema,
  cityId: z.string(),
  address: z.string().max(500).optional(),
  latitude: z.number().min(-90).max(90),
  longitude: z.number().min(-180).max(180),
  ssid: z.string().max(100).optional(),
  requiresPassword: z.boolean(),
  isFree: z.boolean(),
  speedMbps: z.number().min(0).optional(),
  openingHours: z.string().max(200).optional(),
  description: z.string().max(2000).optional(),
  imageUrls: z.array(z.string().url()).optional(),
  poiId: z.string().optional(),
});

const updateSpotBodySchema = z.object({
  name: z.string().min(1).max(200).optional(),
  nameEn: z.string().max(200).optional(),
  type: wifiSpotTypeSchema.optional(),
  address: z.string().max(500).optional(),
  latitude: z.number().min(-90).max(90).optional(),
  longitude: z.number().min(-180).max(180).optional(),
  ssid: z.string().max(100).optional(),
  requiresPassword: z.boolean().optional(),
  isFree: z.boolean().optional(),
  speedMbps: z.number().min(0).optional(),
  openingHours: z.string().max(200).optional(),
  description: z.string().max(2000).optional(),
  imageUrls: z.array(z.string().url()).optional(),
});

const createCredentialBodySchema = z.object({
  wifiSpotId: z.string().optional(),
  name: z.string().min(1).max(200),
  ssid: z.string().min(1).max(100),
  password: z.string().min(1).max(200),
  securityType: securityTypeSchema.optional(),
  locationName: z.string().max(200).optional(),
  latitude: z.number().min(-90).max(90).optional(),
  longitude: z.number().min(-180).max(180).optional(),
  notes: z.string().max(1000).optional(),
  isShared: z.boolean().optional(),
});

const updateCredentialBodySchema = z.object({
  name: z.string().min(1).max(200).optional(),
  ssid: z.string().min(1).max(100).optional(),
  password: z.string().min(1).max(200).optional(),
  securityType: securityTypeSchema.optional(),
  locationName: z.string().max(200).optional(),
  latitude: z.number().min(-90).max(90).optional(),
  longitude: z.number().min(-180).max(180).optional(),
  notes: z.string().max(1000).optional(),
  isShared: z.boolean().optional(),
});

const createReviewBodySchema = z.object({
  wifiSpotId: z.string(),
  speedRating: z.number().min(1).max(5),
  stabilityRating: z.number().min(1).max(5),
  easeOfAccessRating: z.number().min(1).max(5),
  overallRating: z.number().min(1).max(5),
  comment: z.string().max(2000).optional(),
  speedTestResult: z.number().min(0).optional(),
  connectionTime: z.string().max(50).optional(),
  deviceType: z.string().max(50).optional(),
  visitDate: z.string().max(20).optional(),
});

const updateReviewBodySchema = z.object({
  speedRating: z.number().min(1).max(5).optional(),
  stabilityRating: z.number().min(1).max(5).optional(),
  easeOfAccessRating: z.number().min(1).max(5).optional(),
  overallRating: z.number().min(1).max(5).optional(),
  comment: z.string().max(2000).optional(),
  speedTestResult: z.number().min(0).optional(),
});

const listReviewsQuerySchema = z.object({
  limit: z.coerce.number().min(1).max(100).optional(),
  offset: z.coerce.number().min(0).optional(),
});

// ============================================
// Public Routes (no auth required)
// ============================================

export const publicWifiRoutes = new Hono();

// List WiFi spots
publicWifiRoutes.get(
  '/spots',
  zValidator('query', listSpotsQuerySchema),
  async (c) => {
    const { cityId, type, limit } = c.req.valid('query');
    const spots = await WifiSpotService.list(cityId, type, limit);
    return c.json({ data: spots });
  }
);

// Get WiFi spot by ID
publicWifiRoutes.get('/spots/:id', async (c) => {
  const id = c.req.param('id');
  const spot = await WifiSpotService.getById(id);
  return c.json({ data: spot });
});

// Get nearby WiFi spots
publicWifiRoutes.get(
  '/spots/nearby',
  zValidator('query', nearbySpotsQuerySchema),
  async (c) => {
    const query = c.req.valid('query');
    const spots = await WifiSpotService.getNearby(query);
    return c.json({ data: spots });
  }
);

// Search WiFi spots
publicWifiRoutes.get(
  '/spots/search',
  zValidator('query', searchSpotsQuerySchema),
  async (c) => {
    const { query, cityId, type, limit } = c.req.valid('query');
    const spots = await WifiSpotService.search(query, cityId, type, limit);
    return c.json({ data: spots });
  }
);

// List reviews for a WiFi spot
publicWifiRoutes.get(
  '/spots/:spotId/reviews',
  zValidator('query', listReviewsQuerySchema),
  async (c) => {
    const spotId = c.req.param('spotId');
    const { limit, offset } = c.req.valid('query');
    const reviews = await WifiReviewService.listBySpot(spotId, limit, offset);
    return c.json({ data: reviews });
  }
);

// Get shared credentials for a WiFi spot (community passwords)
publicWifiRoutes.get('/spots/:spotId/shared-credentials', async (c) => {
  const spotId = c.req.param('spotId');
  const credentials = await WifiCredentialService.getSharedBySpot(spotId);
  // Mask passwords partially for security
  const masked = credentials.map((cred) => ({
    ...cred,
    password: cred.password.slice(0, 2) + '****' + cred.password.slice(-2),
  }));
  return c.json({ data: masked });
});

// ============================================
// Protected Routes (auth required)
// ============================================

export const wifiRoutes = new Hono<{ Variables: AuthContext }>();

// --- WiFi Spots ---

// Create a new WiFi spot
wifiRoutes.post(
  '/spots',
  zValidator('json', createSpotBodySchema),
  async (c) => {
    const body = c.req.valid('json');
    const userId = c.get('userId');
    const token = c.get('accessToken');

    const id = await WifiSpotService.create(
      { ...body, submittedBy: userId },
      token
    );

    return c.json({ data: { id } }, 201);
  }
);

// Update a WiFi spot
wifiRoutes.patch(
  '/spots/:id',
  zValidator('json', updateSpotBodySchema),
  async (c) => {
    const id = c.req.param('id');
    const body = c.req.valid('json');
    const token = c.get('accessToken');

    const spot = await WifiSpotService.update(id, body, token);
    return c.json({ data: spot });
  }
);

// Delete a WiFi spot
wifiRoutes.delete('/spots/:id', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  await WifiSpotService.remove(id, token);
  return c.json({ success: true });
});

// Verify a WiFi spot (admin action)
wifiRoutes.post('/spots/:id/verify', async (c) => {
  const id = c.req.param('id');
  const userId = c.get('userId');
  const token = c.get('accessToken');

  const spot = await WifiSpotService.verify(id, userId, token);
  return c.json({ data: spot });
});

// --- WiFi Credentials ---

// List user's saved WiFi credentials
wifiRoutes.get('/credentials', async (c) => {
  const userId = c.get('userId');
  const token = c.get('accessToken');
  const limit = c.req.query('limit')
    ? parseInt(c.req.query('limit')!, 10)
    : undefined;

  const credentials = await WifiCredentialService.listByUser(
    userId,
    limit,
    token
  );
  return c.json({ data: credentials });
});

// Get credential by ID
wifiRoutes.get('/credentials/:id', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  const credential = await WifiCredentialService.getById(id, token);
  return c.json({ data: credential });
});

// Get credential for a specific spot
wifiRoutes.get('/spots/:spotId/credential', async (c) => {
  const spotId = c.req.param('spotId');
  const userId = c.get('userId');
  const token = c.get('accessToken');

  const credential = await WifiCredentialService.getBySpot(
    userId,
    spotId,
    token
  );
  return c.json({ data: credential });
});

// Search user's credentials
wifiRoutes.get('/credentials/search', async (c) => {
  const userId = c.get('userId');
  const token = c.get('accessToken');
  const query = c.req.query('query') || '';
  const limit = c.req.query('limit')
    ? parseInt(c.req.query('limit')!, 10)
    : undefined;

  const credentials = await WifiCredentialService.search(
    userId,
    query,
    limit,
    token
  );
  return c.json({ data: credentials });
});

// Create a new WiFi credential
wifiRoutes.post(
  '/credentials',
  zValidator('json', createCredentialBodySchema),
  async (c) => {
    const body = c.req.valid('json');
    const userId = c.get('userId');
    const token = c.get('accessToken');

    const id = await WifiCredentialService.create({ ...body, userId }, token);
    return c.json({ data: { id } }, 201);
  }
);

// Update a WiFi credential
wifiRoutes.patch(
  '/credentials/:id',
  zValidator('json', updateCredentialBodySchema),
  async (c) => {
    const id = c.req.param('id');
    const body = c.req.valid('json');
    const token = c.get('accessToken');

    const credential = await WifiCredentialService.update(id, body, token);
    return c.json({ data: credential });
  }
);

// Mark credential as recently used
wifiRoutes.post('/credentials/:id/mark-used', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  const credential = await WifiCredentialService.markUsed(id, token);
  return c.json({ data: credential });
});

// Delete a WiFi credential
wifiRoutes.delete('/credentials/:id', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  await WifiCredentialService.remove(id, token);
  return c.json({ success: true });
});

// --- WiFi Reviews ---

// List user's reviews
wifiRoutes.get('/reviews', async (c) => {
  const userId = c.get('userId');
  const token = c.get('accessToken');
  const limit = c.req.query('limit')
    ? parseInt(c.req.query('limit')!, 10)
    : undefined;

  const reviews = await WifiReviewService.listByUser(userId, limit, token);
  return c.json({ data: reviews });
});

// Get review by ID
wifiRoutes.get('/reviews/:id', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  const review = await WifiReviewService.getById(id, token);
  return c.json({ data: review });
});

// Get user's review for a specific spot
wifiRoutes.get('/spots/:spotId/my-review', async (c) => {
  const spotId = c.req.param('spotId');
  const userId = c.get('userId');
  const token = c.get('accessToken');

  const review = await WifiReviewService.getUserReview(userId, spotId, token);
  return c.json({ data: review });
});

// Create or update a review
wifiRoutes.post(
  '/reviews',
  zValidator('json', createReviewBodySchema),
  async (c) => {
    const body = c.req.valid('json');
    const userId = c.get('userId');
    const token = c.get('accessToken');

    const id = await WifiReviewService.create({ ...body, userId }, token);
    return c.json({ data: { id } }, 201);
  }
);

// Update a review
wifiRoutes.patch(
  '/reviews/:id',
  zValidator('json', updateReviewBodySchema),
  async (c) => {
    const id = c.req.param('id');
    const body = c.req.valid('json');
    const token = c.get('accessToken');

    const review = await WifiReviewService.update(id, body, token);
    return c.json({ data: review });
  }
);

// Mark a review as helpful (toggle)
wifiRoutes.post('/reviews/:id/helpful', async (c) => {
  const id = c.req.param('id');
  const userId = c.get('userId');
  const token = c.get('accessToken');

  const review = await WifiReviewService.markHelpful(id, userId, token);
  return c.json({ data: review });
});

// Delete a review
wifiRoutes.delete('/reviews/:id', async (c) => {
  const id = c.req.param('id');
  const token = c.get('accessToken');

  await WifiReviewService.remove(id, token);
  return c.json({ success: true });
});
